golem-cli/src/service/component.rs

```diff
@@ -77,6 +90,145 @@ pub trait ComponentService {
 
 pub struct ComponentServiceLive<ProjectContext> {
     pub client: Box<dyn ComponentClient<ProjectContext = ProjectContext> + Send + Sync>,
+    pub file_download_client: Box<dyn FileDownloadClient + Send + Sync>,
+}
+
+impl<ProjectContext> ComponentServiceLive<ProjectContext> {
+    async fn load_file(
+        &self,
+        component_file: InitialComponentFile,
+    ) -> Result<Vec<LoadedFile>, GolemError> {
+        let scheme = component_file.source_path.as_url().scheme();
+        match scheme {
+            "file" | "" => self.load_local_file(component_file).await,
+            "http" | "https" => self
+                .download_remote_file(component_file)
+                .await
+                .map(|f| vec![f]),
+            _ => Err(GolemError(format!("Unsupported scheme: {}", scheme))),
+        }
+    }
+
+    async fn load_local_file(
+        &self,
+        component_file: InitialComponentFile,
+    ) -> Result<Vec<LoadedFile>, GolemError> {
+        // if it's a directory, we need to recursively load all files and combine them with their target paths and permissions.
+        let source_path = PathBuf::from(component_file.source_path.as_url().path());
+
+        let mut results: Vec<LoadedFile> = vec![];
+        let mut queue: VecDeque<(PathBuf, ComponentFilePathAndPermissions)> =
+            vec![(source_path, component_file.target)].into();
+
+        while let Some((path, target)) = queue.pop_front() {
+            if path.is_dir() {
+                let read_dir = tokio::fs::read_dir(&path).await.map_err(|err| {
+                    GolemError(format!("Error reading directory {}: {err}", path.display()))
+                })?;
+                let mut read_dir_stream = ReadDirStream::new(read_dir);
+
+                while let Some(entry) = read_dir_stream.next().await {
+                    let entry = entry.map_err(|err| {
+                        GolemError(format!("Error reading directory entry: {}", err))
+                    })?;
+                    let next_path = entry.path();
+
+                    let file_name = entry.file_name().into_string().map_err(|_| {
+                        GolemError("Error converting file name to string".to_string())
```

Comment:
May be it's a good idea to use the actual error. This is because, given it's a corner case to fail, it would be good to know, what exactly made it fail.

Reply:
Docs for into_string say
```
Converts the OsString into a String  if it contains valid Unicode data.

On failure, ownership of the original OsString is returned.
```

So in the case of error we would not get an actual error only a string which we cannot even display without conversion first. Adjusted the error to better reflect this.

Reply:
Sure. Makes sense.